server/src/main/java/org/gluu/oxtrust/action/Authenticator.java

```diff
@@ -421,4 +462,153 @@ private String getOxAuthHost(String oxAuthAuthorizeUrl) {
 		return null;
 	}
 
+	private WebTarget getNormalAuthenticationRequest(WebTarget target, String clientId, String nonce,
+													 String state, String acrValues, String scope) {
+		String responseType = "code";
+
+		target = target.queryParam(OxTrustConstants.OXAUTH_CLIENT_ID, clientId);
+		target = target.queryParam(OxTrustConstants.OXAUTH_REDIRECT_URI, appConfiguration.getLoginRedirectUrl());
+		target = target.queryParam(OxTrustConstants.OXAUTH_RESPONSE_TYPE, responseType);
+		target = target.queryParam(OxTrustConstants.OXAUTH_SCOPE, scope);
+		target = target.queryParam(OxTrustConstants.OXAUTH_NONCE, nonce);
+		target = target.queryParam(OxTrustConstants.OXAUTH_STATE, state);
+		if (StringHelper.isNotEmpty(acrValues)) {
+			target = target.queryParam(OxTrustConstants.OXAUTH_ACR_VALUES, acrValues);
+		}
+		return target;
+	}
+
+	private WebTarget getFapiAuthenticationRequest(WebTarget target, String acrValues, String nonce,
+												   String state, String clientId, String scope) {
+		try {
+			String responseType = "code id_token";
+
+			Jwt jwt = new Jwt();
+			JwtSubClaimObject acr = new JwtSubClaimObject();
+			acr.setClaim("value", acrValues);
+			acr.setClaim("essential", true);
+
+			JwtSubClaimObject idToken = new JwtSubClaimObject();
+			idToken.setClaim("acr", acr.toJsonObject());
+
+			JwtSubClaimObject claims = new JwtSubClaimObject();
+			claims.setClaim("id_token", idToken.toJsonObject());
+
+			JwtClaims jwtClaims = new JwtClaims();
+			jwtClaims.setClaim("aud", appConfiguration.getOxAuthIssuer());
+			jwtClaims.setClaim("scope", StringUtils.replace(scope, "+", " "));
+			jwtClaims.setClaim("claims", claims.toJsonObject());
+			jwtClaims.setClaim("iss", clientId);
+			jwtClaims.setClaim("response_type", responseType);
+			jwtClaims.setClaim("redirect_uri", appConfiguration.getLoginRedirectUrl());
+			jwtClaims.setClaim("state", state);
+			jwtClaims.setClaim("exp", new Date().getTime() / 1000 + 180L);
+			jwtClaims.setClaim("nonce", nonce);
+			jwtClaims.setClaim("client_id", clientId);
+
+			jwt.setClaims(jwtClaims);
+
+			JwtHeader jwtHeader = new JwtHeader();
+			jwtHeader.setAlgorithm(SignatureAlgorithm.PS256);
+			jwtHeader.setKeyId(clientId);
+			jwt.setHeader(jwtHeader);
+
+			RSAPrivateKey privateKey = new RSAPrivateKey(appConfiguration.getOxAuthClientFapiJWKPrivateKeyModulus(),
+					appConfiguration.getOxAuthClientFapiJWKPrivateKeyPrivateExponent());
```

Comment:
Again, it's not global configuration. It's private key of the client. Please look at client's `oxAuthJwks` and `oxAuthJwksURI`. RP (here oxtrust) should load key from own keystore.
```
PrivateKey key = (PrivateKey) keyStore.getKey(alias, keyStoreSecret.toCharArray());
```

Reply:
Ok, I think have that configuration in a keystore would be better.
Question, should I filter `oxAuthJwks` list? because this is a list and we could have many values, if so, which criterias should I use? perhaps `use=sig`? or should I put a new oxTrust json property to select a key, maybe `kid` property.

Reply:
You should pick up kid by algorithm which is set in `oxAuthTokenEndpointAuthSigningAlg`. No new properties.

Reply:
Ok, I think I have some incorrect configuration, because now I have `oxAuthTokenEndpointAuthSigningAlg=ES256` and the unique key that I have is `PS256`. Anyway, I understand your point, just filter the list of keys in `oxAuthJwks` and use the corresponding key.